app/src/main/java/com/cdio/solitaire/model/CardStack.kt

```
@@ -0,0 +1,144 @@
+
+    /**
+     * Standard push, pushing a cardStack to the tail of the stack
+     */
+    fun pushStack(stack: CardStack) { // Push a CardStack to tail
+        if (stack.size == 0) return
+        var card = stack.head
+        for (i in 1..stack.size) {
+            card!!.stackID = stackID
+            card = card.next
+        }
+
+        when (size) {
+            0 -> {
+                head = stack.head
+                tail = stack.tail
+            }
+            else -> {
+                tail!!.next = stack.head
+                stack.head!!.prev = tail
+                tail = stack.tail
+            }
+        }
+        size += stack.size
+        stack.resetCardStack()
+    }
+
+    /**
+     * Standard pop, popping a stack off the tail of the stack.
+     */
+    fun popStack(cardsToPop: Int): CardStack? {
+        if (size == 0 || size < cardsToPop) return null
+
+        val poppedStack = CardStack(-1)
+        if (size - cardsToPop == 0) poppedStack.pushStack(this)
+        else {
+            poppedStack.tail = tail
+            poppedStack.head = tail
+            // TODO: The following for loop is NOT needed if you already know the address of
+            //  poppedStack.head, changing this will make this O(1), instead of O(n)
+            for (i in 1 until cardsToPop) poppedStack.head = poppedStack.head!!.prev
+            tail = poppedStack.head!!.prev
+            tail!!.next = null
+            poppedStack.head!!.prev = null
+            poppedStack.size = cardsToPop
+            size -= cardsToPop
+        }
+        return poppedStack
+    }
+
+    /**
+     * Special push used for stock specifically when flipping talon. This pushes the stack to head
+     * instead of tail.
+     */
+    fun pushStackToHead(stack: CardStack) {
+        if (stack.size == 0) {
+            Log.e(
+                "EmptyStackPush",
+                "Trying to push empty stack, stackID: " + stack.stackID.toString()
+            )
+            return
```

Consider throwing an exception here instead of just logging the error, since a stack trace maybe could provide a lot of useful information.

That also means the method doesn't have to return a nullable object, which can make the code simpler.